gh-127266: avoid data races when updating type slots v2

nascheme opened the pull request:
This is an updated version of GH-131174, which was reverted.  I figured the cleanest thing to do is make a new PR.

This is the same as the previous PR with the following additional change. The `update_all_slots()` and `type_setattro()` functions are now more careful when the world is stopped.  Instead of doing the MRO lookups while the world is stopped, we do them all first and collect the slot pointers to be updated.  Then, we stop the world and do those updates.  This makes it much easier to confirm the code running during the stop-the-world is safe and that should avoid the deadlocks.

The `test_opcache` test has become quite a bit slower.  It seems to be due to mutex contention in the `__getitem__` and `__getattribute__` method assignment tests.   I reduced the items count from 1000 to 100 to keep the test from becoming much slower.


<!-- gh-issue-number: gh-127266 -->
* Issue: gh-127266
<!-- /gh-issue-number -->

nascheme committed:
gh-127266: avoid data races when updating type slots

In the free-threaded build, avoid data races caused by updating type
slots or type flags after the type was initially created.  For those
(typically rare) cases, use the stop-the-world mechanism.  Remove the
use of atomics when reading or writing type flags.  The use of atomics
is not sufficient to avoid races (since flags are sometimes read without
a lock and without atomics) and are no longer required.

nascheme committed:
For update_all_slots(), do updates more safely.

To avoid deadlocks while the world is stopped, we need to avoid calling APIs
like _PyObject_HashFast() and _PyDict_GetItemRef_KnownHash().  Collect the
slot updates to be done and then apply them all at once.  This reduces the
amount of code running in the stop-the-world condition.

nascheme committed:
Avoid "empty structure" compile error.

nascheme committed:
Use apply_slot_updates() for type_setattro().

nascheme committed:
Merge 'origin/main' into type-slot-ts-v2

nascheme committed:
Reduce number of items in test for slot updates.

Now that stop-the-world is used to do the slot update, these tests
are a lot slower in the free-threaded build.  Test with fewer items,
which will still hopefully be enough to find bugs in the specializer.

nascheme committed:
Add TSAN suppression for _Py_slot_tp_getattr_hook.

nascheme committed:
Queue update of tp_flags as well.

The clearing of Py_TPFLAGS_HAVE_VECTORCALL must be done when the world
is stopped too.

nascheme committed:
Performance, skip stop-the-world when possible.

Since we stack allocate one chunk, we need to check 'n' to see if there
are actually any updates to make.  It's pretty common that no updates
are actually needed.

nascheme committed:
Merge 'origin/main' into type-slot-ts-v2

nascheme committed:
Always clear version after __bases__ update.

nascheme committed:
Merge 'origin/main' into type-slot-ts-v2

nascheme committed:
Add test for assigning __bases__.

colesbury commented on Lib/test/test_opcache.py:
```diff
@@ -576,6 +576,7 @@ class TestRacesDoNotCrash(TestBase):
     # Careful with these. Bigger numbers have a higher chance of catching bugs,
     # but you can also burn through a *ton* of type/dict/function versions:
     ITEMS = 1000
+    SMALL_ITEMS = 100
```
It might be worth investigating this further. I'm surprised there's a large slowdown in this PR, but not in the earlier version. What's the relevant difference?

nascheme replied:
I timed the most recent version of this PR vs "main", using `SMALL_ITEMS = 1000`.  It seems the slowdown is not too bad now.  I've left it at 100 but we could revert to 1000 if we want to.

I wasn't able to exactly pin-point the reason for the huge slowdown before.  Based on "samply" profiling, it looked like there was a lot of contention for the `TYPE_LOCK` mutex and the STM mutex.  Quite a lot of time was spent in `_Py_yield()`.   That's kind of expected since the two writers are both trying to acquire both of those.

nascheme replied:
I reverted this change to test_opcache.

colesbury commented on Tools/tsan/suppressions_free_threading.txt:
```diff
@@ -46,4 +40,7 @@ race:list_inplace_repeat_lock_held
+race:PyObject_Realloc
+
+# This function writes to the tp_getattr slot in an unsafe way.
+race:_Py_slot_tp_getattr_hook
```
Hmm.. seems like we should fix this in a follow up PR

nascheme replied:
I created an issue for this and the couple other unsafe stores.  This PR doesn't fix everything, unfortunately.

colesbury commented on Objects/typeobject.c:
```diff
@@ -1759,20 +1819,29 @@ type_set_bases_unlocked(PyTypeObject *type, PyObject *new_bases)
 
     set_tp_bases(type, Py_NewRef(new_bases), 0);
-    type->tp_base = (PyTypeObject *)Py_NewRef(new_base);
+    type->tp_base = (PyTypeObject *)Py_NewRef(best_base);
```
Should we be setting `type->tp_base` in a stop the world pause?

Doesn't have to be part of this PR if it complicates things.

nascheme replied:
Yeah, probably should be.  I'm going to leave it for a followup.

nascheme committed:
Avoid releasing TYPE_LOCK when stopping the world.

nascheme committed:
Merge 'origin/main' into type-slot-ts-v2

nascheme committed:
Add issue number for TSAN suppression.

nascheme committed:
Bug fix for type_lock_prevent_release().

If the two mutex form of the critical section is used, need to put the
other mutex into '_cs_mutex'.

nascheme committed:
Merge 'origin/main' into type-slot-ts-v2

kumaraditya303 commented on Objects/typeobject.c:
```diff
@@ -1075,7 +1177,7 @@ type_modified_unlocked(PyTypeObject *type)
        needed.
      */
-    ASSERT_TYPE_LOCK_HELD();
+    ASSERT_NEW_TYPE_OR_LOCKED(type);
```
```suggestion
    ASSERT_NEW_TYPE_OR_LOCKED(type);
    assert(!types_world_is_stopped());
```
Since this is re-entrant, the world shouldn't be stopped.

nascheme committed:
Add additional assert.

nascheme committed:
Merge 'origin/main' into type-slot-ts-v2

nascheme committed:
Revert test_opcache item size change.

This test gets a bit slower, due to stop-the-world but it is not so
dramatic.

nascheme committed:
Add comment for new unit test.

nascheme committed:
Fix assert for default build.

colesbury commented on Objects/typeobject.c:
```diff
@@ -3608,13 +3726,167 @@ solid_base(PyTypeObject *type)
+}
+
+static void
+apply_slot_updates_world_stopped(slot_update_t *updates)
```
The naming convention here isn't consistent with what we use for locks: we use "_lock_held" for functions that should be called with the lock already held. I would think that "_world_stopped" would mean that the function should be called with the world already stopped (instead of stopping the world internally).

nascheme committed:
Improve function name.

The apply_slot_updates_world_stopped() name implies that the world
is already stopped, based on convention.  Rename for clarity.
